scipy/stats/_hypotests.py

```diff
@@ -9,6 +9,276 @@
 from ._continuous_distns import chi2, norm
 from scipy.special import gamma, kv, gammaln
 from . import _wilcoxon_data
+import scipy.stats.stats
+from functools import wraps
+from scipy._lib._docscrape import FunctionDoc, Parameter
```

Comment:
I used it to add the new parameters `axis` and `nan_policy` to the documentation of functions wrapped by the decorator.

No, I copied the whole file as it seems useful. It's not that much longer than what I used. If i remember (can't check easily from my phone now) I didn't use the parts that are specific to class documentation.

Reply:
ok I guess it's fine. A bit annoying that we have to copy this over as just in `numpydoc`. You could not easily arrive to the same output using the same way as here?

https://github.com/scipy/scipy/blob/d1a683f5716e50ef84a61473518c15deb4a4d9f7/scipy/stats/_distn_infrastructure.py#L833

Reply:
I'm not familiar with that function, but its purpose seems orthogonal to the stuff in the `_docscrape`.py file I added. The `_docscrape` stuff I'm adding parses an existing docstring. `_construct_default_doc` seems to generate documentation for stats distributions.

The `_docscrape` stuff is very useful. For instance, we could use it in `_construct_default_doc` to avoid the need to add strings like `(before_notes)s` in the docstring and perform substitution. Instead, we would use the `_docscrape` stuff to parse the docstring and decide where the before notes and after notes go.

The overall approach to the stats distribution docstrings didn't seem like a golden example to follow, so I didn't try very hard to figure it out. It's not immediately obvious to me how it all works, or why the documentation you get when you type `help(stats.norm)` in the console is so different from the rendered `stats.norm` documentation. 

Anyway, we could avoid copying the `_docscrape.py` stuff. We'd eliminate:
```
        doc = FunctionDoc(vectorize_hypotest_wrapper)
        parameter_names = [param.name for param in doc['Parameters']]
        if 'axis' in parameter_names:
            doc['Parameters'][parameter_names.index('axis')] = (
                _axis_parameter_doc)
        else:
            doc['Parameters'].append(_axis_parameter_doc)
        if 'nan_policy' in parameter_names:
            doc['Parameters'][parameter_names.index('nan_policy')] = (
                _nan_policy_parameter_doc)
        else:
            doc['Parameters'].append(_nan_policy_parameter_doc)
        doc = str(doc).split("\n", 1)[1]  # remove signature
```
which takes apart the docstring, adds the axis and nan_policy documentation to the right place, and rewrites the docstring. Instead, we'd we'd have to manually add tags like `<nan_policy_documentation>` and `<axis_documentation>` to the docstrings of every function to which the decorator is applied, and the code above would be replaced with code that substitutes the actual parameter documentation in for these tags.

I think it's better to copy the file than to have to muddy the docstrings with these tags.

Reply:
>  it's really not pretty and also very specialized. So if we can not follow this and have something cleaner to rely on, then great

Oops, I'm not sure I understand. Which code would you rather not rely on?
Is it OK to keep `_lib/_docscrape.py` and use it as I have?

Reply:
Sorry I was referring to the function in `scipy/scipy/stats/_distn_infrastructure.py`. I agree with you that taking this file directly import from `numpydoc` is a good option.